openforcefield/topology/molecule.py

```diff
@@ -0,0 +1,3372 @@
+            
+    @property
+    def molecule_atom_index(self):
+        """
+        The index of this Atom within the the list of atoms in ``Molecules``.
+        Note that this can be different from ``particle_index``.
+
+        """
+        if self._molecule is None:
+            raise ValueError('This Atom does not belong to a Molecule object')
+        return self._molecule.atoms.index(self)
+
+
+
+    ## From Jeff: Not sure if we actually need this
+    @property
+    def topology_atom_index(self):
+        """
+        The index of this Atom within the the list of atoms in ``Topology``.
+        Note that this can be different from ``particle_index``.
+
+        """
+        if self._topology is None:
+            raise ValueError('This Atom does not belong to a Topology object')
+        # TODO: This will be slow; can we cache this and update it only when needed?
+        #       Deleting atoms/molecules in the Topology would have to invalidate the cached index.
+        return self._topology.atoms.index(self)
+
+    def __repr__(self):
+        # TODO: Also include particle_index and which molecule this atom belongs to?
+        return "Atom(name={}, atomic number={})".format(self._name, self._atomic_number)
+
+    def __str__(self):
+        # TODO: Also include particle_index and which molecule this atom belongs to?
+        return "<Atom name='{}' atomic number='{}'>".format(self._name, self._atomic_number)
+
+#=============================================================================================
+# VirtualSite
+#=============================================================================================
+
+class VirtualSite(Particle):
+    """
+    A particle representing a virtual site whose position is defined in terms of ``Atom`` positions.
+
+    .. warning :: This API experimental and subject to change.
+
+    Note that chemical atoms are represented by the ``Atom``.
+
+    .. todo::
+
+       * Should a virtual site be able to belong to more than one Topology?
+       * Should virtual sites be immutable or mutable?
+
+    """
+
+    def __init__(self, atoms, charge_increments=None, epsilon=None, sigma=None, rmin_half=None, name=None):
+        """
+        Base class for VirtualSites
+
+        .. todo ::
+
+           * This will need to be generalized for virtual sites to allow out-of-plane sites, which are not simply a linear combination of atomic positions
+           * Add checking for allowed virtual site types
+           * How do we want users to specify virtual site types?
+
+        Parameters
+        ----------
+        atoms : list of Atom of shape [N]
+            atoms[index] is the corresponding Atom for weights[index]
+        charge_increments : list of floats of shape [N], optional, default=None
+            The amount of charge to remove from the VirtualSite's atoms and put in the VirtualSite. Indexing in this list should match the ordering in the atoms list. Default is None.
```

Comment:
It could help readability to split long comments like this across multiple lines.